src/components/data/SavedColumnSettings.js

```diff
@@ -0,0 +1,280 @@
+  _.chunk(2),
+  _.map(([name, visible]) => ({ name, visible }))
+)
+
+const useSavedColumnSettings = ({ workspaceId, snapshotName, entityMetadata, entityType }) => {
+  // Saved column settings for all tables are stored in the same attribute in the format:
+  // {
+  //   snapshots: {
+  //      snapshot_name: {
+  //        table_name: {
+  //          settings_name: settings,
+  //          ...
+  //        },
+  //        ...
+  //      },
+  //      ...
+  //    },
+  //    tables: {
+  //      table_name: {
+  //        settings_name: settings,
+  //        ...
+  //      },
+  //      ...
+  //    }
+  // }
+  const baseKey = snapshotName ? `snapshots.${snapshotName}` : 'tables'
+  const entityTypeKey = `${baseKey}.${entityType}`
+
+  const signal = useCancellation()
+
+  const getAllSavedColumnSettings = async () => {
+    const { namespace, name } = workspaceId
+    const { workspace: { attributes } } = await Ajax(signal).Workspaces.workspace(namespace, name).details(['workspace.attributes'])
+    return _.flow(
+      _.getOr('{}', savedColumnSettingsWorkspaceAttributeName),
+      JSON.parse,
+      // Drop any column settings for entity types that no longer exist
+      _.update(baseKey, _.pick(_.keys(entityMetadata)))
+    )(attributes)
+  }
+
+  const updateSavedColumnSettings = async allColumnSettings => {
+    const { namespace, name } = workspaceId
+    await Ajax(signal).Workspaces.workspace(namespace, name).shallowMergeNewAttributes({
+      // Store settings as a string since Rawls expects lists of objects to be entity references
+      [savedColumnSettingsWorkspaceAttributeName]: JSON.stringify(allColumnSettings)
+    })
+  }
+
+  const getSavedColumnSettings = async () => {
+    const allSavedColumnSettings = await getAllSavedColumnSettings()
+    const columnSettingsForEntityType = _.getOr({}, entityTypeKey, allSavedColumnSettings)
+    const entityTypeAttributes = entityMetadata[entityType].attributeNames
+
+    return _.flow(
+      _.mapValues(decodeColumnSettings),
+      // Reconcile saved settings with any changes to entity type attributes
+      _.mapValues(columnSettings => {
+        return _.concat(
+          // Remove columns that no longer exist on the entity type from settings
+          _.filter(({ name }) => _.includes(name, entityTypeAttributes), columnSettings),
+          // Add columns that were not in the saved settings
+          _.flow(
+            _.without(_.map(_.get('name'), columnSettings)),
+            _.map(name => ({ name, visible: false }))
+          )(entityTypeAttributes)
+        )
+      }),
+      // Remove any settings that do not contain any existing columns
+      _.pickBy(columnSettings => _.size(columnSettings) > 0)
+    )(columnSettingsForEntityType)
+  }
+
+  const saveColumnSettings = async (columnSettingsName, columnSettings) => {
+    // Re-fetch column settings to avoid overwriting changes made by other users since we first loaded settings.
```

Comment:
This is great, but it only reduces the likelihood of a race rather than eliminate it. I would love to see a ticket to track fixing this at the API layer.